dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java

```diff
@@ -738,114 +737,87 @@ private int startFetch()
+                catch(Exception ie){}
+                return -1;
             }
-            catch (IOException e)
+
+            // If len < 0 we have finished reading the file
+            if (response.len < 0)
             {
-                prefetchException = new HpccFileException(e.getMessage());
+                try
+                {
+                    close();
+                }
+                catch (IOException e)
+                {
+                    prefetchException = new HpccFileException(e.getMessage());
+                }
+                return -1;
             }
-            return -1;
-        }
 
-        // Invalid response len, should be at least 4
-        if (len < 4)
-        {
-            prefetchException = new HpccFileException("Early data termination, no handle");
-            try
+            if (response.errorCode != RFCCodes.RFCStreamNoError)
             {
-                close();
+                prefetchException = new HpccFileException(response.errorMessage);
+                return -1;
             }
-            catch(Exception e){}
-            return -1;
-        }
 
-        //------------------------------------------------------------------------------
-        // Read the new handle, and retry with the token if it is invalid
-        //------------------------------------------------------------------------------
+            //------------------------------------------------------------------------------
+            // Retry with the token if handle is invalid
+            //------------------------------------------------------------------------------
 
-        try
-        {
-            this.handle = dis.readInt();
-            if (this.handle == 0)
+            this.handle = response.handle;
+            if (this.handle == 0 && !inTokenRetry)
             {
+                inTokenRetry = true;
+
+                String retryTrans = this.makeTokenRequest();
+                int len = retryTrans.length();
                 try
                 {
-                    len = retryWithToken();
-                    log.warn("Unable to make request with handle, retyring with token.");
+                    this.dos.writeInt(len);
+                    this.dos.write(retryTrans.getBytes(HPCCCharSet), 0, len);
+                    this.dos.flush();
                 }
-                catch (HpccFileException e)
+                catch (IOException e)
                 {
-                    prefetchException = e;
-                    try
-                    {
-                        close();
-                    }
-                    catch(Exception ie){}
+                    prefetchException = new HpccFileException("Failed on remote read read retry", e);
                     return -1;
                 }
-
-                if (len == 0)
+            }
+            else if (this.handle == 0)
+            {
+                prefetchException = new HpccFileException("Read retry failed");
```

Comment:
The if statement above will be executed if we are in a retry, so that exception will only be thrown during a retry.

Reply:
the else-if clause doesn't seem to be considered a retry (based on variables used, and your reply, but the error message in this clause says it is a failed retry). If the error message is correct, ignore, otherwise ensure the message is not misleading.

Reply:
My original comment had a typo. I meant to say the "the if statement above will be executed if we aren't in a retry". So, the error message is correct. If we head a read failure we would take the first branch and move into the token retry, if that failed we would end up in the second branch.

` if (this.handle <= 0 && !inTokenRetry)`